R/data.R

```diff
@@ -204,27 +204,71 @@ any_zero_dims <- function(data) {
+#'
 #' @return Paths to CSV files (one per chain).
 #'
-draws_to_csv <- function(draws, sampler_diagnostics = NULL) {
+#' @details
+#' `draws_to_csv()` generates a CSV suitable for running standalone generated
+#' quantities with CmdStan. The CSV file contains a single comment `#num_samples`,
+#' which equals the number of iterations in the supplied draws object.
+#'
+#' The comment is followed by the column names. The first column is the `lp__` value,
+#' followed by sampler diagnostics and finnaly other variables of the draws object.
+#' #' If the draws object does not contain the `lp__` or sampler diagnostics variables,
+#' columns with zeros are created in order to conform with the requirements of the
+#' standalone generated quantities method of CmdStan.
+#'
+#' The column names line is finally followed by the values of the draws in the same
+#' order as the column names.
+#'
+#' @examples
+#' \dontrun{
+#' draws <- posterior::example_draws()
+#'
+#' draws_csv_files <- draws_to_csv(draws)
+#' }
+#'
+draws_to_csv <- function(draws,
+                         sampler_diagnostics = NULL,
+                         dir = tempdir(),
+                         basename = "fittedParams") {
+  sampler_diagnostics_names <- c(
+    "accept_stat__", "stepsize__", "treedepth__",
+    "n_leapfrog__", "divergent__", "energy__"
+  )
   n <- posterior::niterations(draws)
   n_chains <- posterior::nchains(draws)
+  draws_variables <- posterior::variables(draws)
+  sampler_diagnostics_variables <- posterior::variables(sampler_diagnostics)
+
+  # create dummy sampler diagnostics due to CmdStan requirement for all columns in GQ if needed
   zeros <- rep(0, n * n_chains) # filler for creating dummy sampler diagnostics and lp__ if necessary
   if (is.null(sampler_diagnostics)) {
-    # create dummy sampler diagnostics due to CmdStan requirement for all columns in GQ
-    sampler_diagnostics <- posterior::draws_array(
-      accept_stat__ = zeros,
-      stepsize__ = zeros,
-      treedepth__ = zeros,
-      n_leapfrog__ = zeros,
-      divergent__ = zeros,
-      energy__ = zeros,
-      .nchains = n_chains
-    )
+    missing_sampler_diagnostics <- sampler_diagnostics_names[!(sampler_diagnostics_names %in% draws_variables)]
+
+  } else {
+    missing_sampler_diagnostics <- sampler_diagnostics_names[!(sampler_diagnostics_names %in% draws_variables)]
+    missing_sampler_diagnostics <- missing_sampler_diagnostics[!(missing_sampler_diagnostics %in% draws_variables)]
+  }
+  if (length(missing_sampler_diagnostics) > 0) {
+    additional_sampler_diagnostics <- list()
+    for (name in missing_sampler_diagnostics) {
+      additional_sampler_diagnostics[[name]] <- zeros
+    }
+    additional_sampler_diagnostics[[".nchains"]] <- n_chains
+    additional_sampler_diagnostics <- do.call(posterior::draws_array, additional_sampler_diagnostics)
+    sampler_diagnostics <- posterior::bind_draws(sampler_diagnostics, additional_sampler_diagnostics)
```

Comment:
These changes here are so that this function still works even if all sampler diagnostics columns/variables are part of the draws input of this function and not the separate sampler_diagnostics.